### ISSUE TITLE
Crash with unclean shutdown

### ISSUE THREAD

**Issue by rldleblanc** [kyuupichan/electrumx#637]
My server has been crashing (I think the power supply is going bad). Now ElectrumX rarely starts up properly. I wiped the database files for three coins and two rebuilt fine (BCH was still building during this last crash). I was about 4 minor version back and it just looked like ElectrumX was just hanging. I installed the latest 1.8.11 and even master today and I started getting backtraces (that's when I wiped the db of the three offending coins. I now have LTC (again) and FTC (again) that are broken. DASH that I rebuilt earlier is fine right now whereas FTC was fine until this last crash. BTC so far has not broken.

The files are on BTRFS, so if the low power conditions are corrupting files as they are being written to disk, it should be caught by the checksum. I'm not seeing any checksum errors from BTRFS though.

Is there a way to rewind X blocks instead of going through the whole chain again?

Here is the log from FTC:

Nov 09 22:26:05 electrumx systemd[1]: Started Electrumx FTC.
Nov 09 22:26:06 electrumx python3.7[202]: INFO:Controller:Python version: 3.7.0 (default, Jun 27 2018, 14:40:03)
Nov 09 22:26:06 electrumx python3.7[202]: [GCC 8.1.0]
Nov 09 22:26:06 electrumx python3.7[202]: INFO:Controller:software version: ElectrumX 1.8.11
Nov 09 22:26:06 electrumx python3.7[202]: INFO:Controller:aiorpcX version: 0.10.1
Nov 09 22:26:06 electrumx python3.7[202]: INFO:Controller:supported protocol versions: 1.2-1.4
Nov 09 22:26:06 electrumx python3.7[202]: INFO:Controller:event loop policy: None
Nov 09 22:26:06 electrumx python3.7[202]: INFO:Controller:reorg limit is 2,000 blocks
Nov 09 22:26:06 electrumx python3.7[202]: INFO:Daemon:daemon #1 at 192.168.21.23:19337/ (current)
Nov 09 22:26:06 electrumx python3.7[202]: INFO:DB:switching current directory to /home/rleblanc/electrumx_db/ftc/
Nov 09 22:26:06 electrumx python3.7[202]: INFO:DB:using leveldb for DB backend
Nov 09 22:26:06 electrumx python3.7[202]: ERROR:Daemon:starting up checking blocks.  Retrying occasionally...
Nov 09 22:27:09 electrumx python3.7[202]: ERROR:Daemon:starting up checking blocks.  Retrying occasionally...
Nov 09 22:27:22 electrumx python3.7[202]: INFO:Daemon:running normally
Nov 09 22:27:22 electrumx python3.7[202]: INFO:DB:opened UTXO DB (for sync: True)
Nov 09 22:27:22 electrumx python3.7[202]: INFO:DB:DB version: 6
Nov 09 22:27:22 electrumx python3.7[202]: INFO:DB:coin: Feathercoin
Nov 09 22:27:22 electrumx python3.7[202]: INFO:DB:network: mainnet
Nov 09 22:27:22 electrumx python3.7[202]: INFO:DB:height: 2,406,365
Nov 09 22:27:22 electrumx python3.7[202]: INFO:DB:tip: 430690c2ec2b73793b47b6a58e0e4690493f21870699944c95dd827f403e7121
Nov 09 22:27:22 electrumx python3.7[202]: INFO:DB:tx count: 4,605,127
Nov 09 22:27:22 electrumx python3.7[202]: INFO:DB:flushing DB cache at 1,200 MB
Nov 09 22:27:22 electrumx python3.7[202]: INFO:History:history DB version: 0
Nov 09 22:27:22 electrumx python3.7[202]: INFO:History:flush count: 34,373
Nov 09 22:27:22 electrumx python3.7[202]: INFO:DB:deleted 295 stale undo entries
Nov 09 22:27:22 electrumx python3.7[202]: INFO:DB:deleted 1 stale block files
Nov 09 22:27:22 electrumx python3.7[202]: ERROR:root:task crashed: <Task finished coro=<Controller.serve() done, defined at /usr/local/lib/python3.7/dist-packages/electrumX-1.8.11-py3.7.egg/electrumx/server/controller.py:81> exception=AssertionError()
Nov 09 22:27:22 electrumx python3.7[202]: Traceback (most recent call last):
Nov 09 22:27:22 electrumx python3.7[202]:   File "/usr/local/lib/python3.7/dist-packages/aiorpcX-0.10.1-py3.7.egg/aiorpcx/util.py", line 118, in check_task
Nov 09 22:27:22 electrumx python3.7[202]:     task.result()
Nov 09 22:27:22 electrumx python3.7[202]:   File "/usr/local/lib/python3.7/dist-packages/electrumX-1.8.11-py3.7.egg/electrumx/server/controller.py", line 131, in serve
Nov 09 22:27:22 electrumx python3.7[202]:     await group.spawn(wait_for_catchup())
Nov 09 22:27:22 electrumx python3.7[202]:   File "/usr/local/lib/python3.7/dist-packages/aiorpcX-0.10.1-py3.7.egg/aiorpcx/curio.py", line 242, in __aexit__
Nov 09 22:27:22 electrumx python3.7[202]:     await self.join(wait=self._wait)
Nov 09 22:27:22 electrumx python3.7[202]:   File "/usr/local/lib/python3.7/dist-packages/aiorpcX-0.10.1-py3.7.egg/aiorpcx/curio.py", line 213, in join
Nov 09 22:27:22 electrumx python3.7[202]:     raise task.exception()
Nov 09 22:27:22 electrumx python3.7[202]:   File "/usr/local/lib/python3.7/dist-packages/electrumX-1.8.11-py3.7.egg/electrumx/server/block_processor.py", line 654, in fetch_and_process_blocks
Nov 09 22:27:22 electrumx python3.7[202]:     await self._first_open_dbs()
Nov 09 22:27:22 electrumx python3.7[202]:   File "/usr/local/lib/python3.7/dist-packages/electrumX-1.8.11-py3.7.egg/electrumx/server/block_processor.py", line 634, in _first_open_dbs
Nov 09 22:27:22 electrumx python3.7[202]:     await self.db.open_for_sync()
Nov 09 22:27:22 electrumx python3.7[202]:   File "/usr/local/lib/python3.7/dist-packages/electrumX-1.8.11-py3.7.egg/electrumx/server/db.py", line 146, in open_for_sync
Nov 09 22:27:22 electrumx python3.7[202]:     await self._open_dbs(True, False)
Nov 09 22:27:22 electrumx python3.7[202]:   File "/usr/local/lib/python3.7/dist-packages/electrumX-1.8.11-py3.7.egg/electrumx/server/db.py", line 134, in _open_dbs
Nov 09 22:27:22 electrumx python3.7[202]:     await self._read_tx_counts()
Nov 09 22:27:22 electrumx python3.7[202]:   File "/usr/local/lib/python3.7/dist-packages/electrumX-1.8.11-py3.7.egg/electrumx/server/db.py", line 102, in _read_tx_counts
Nov 09 22:27:22 electrumx python3.7[202]:     assert len(tx_counts) == size
Nov 09 22:27:22 electrumx python3.7[202]: AssertionError
Nov 09 22:30:16 electrumx systemd[1]: Stopping Electrumx FTC...
Nov 09 22:30:16 electrumx python3.7[202]: WARNING:Controller:received SIGTERM signal, initiating shutdown
Nov 09 22:30:16 electrumx python3.7[202]: INFO:Controller:shutting down
Nov 09 22:30:16 electrumx python3.7[202]: INFO:Controller:shutdown complete
Nov 09 22:30:16 electrumx systemd[1]: Stopped Electrumx FTC.
Nov 09 22:30:16 electrumx systemd[1]: electrumx_ftc.service: Failed to reset devices.list: Operation not permitted
Nov 09 22:30:16 electrumx systemd[1]: Started Electrumx FTC.
Nov 09 22:30:16 electrumx python3.7[548]: INFO:Controller:Python version: 3.7.0 (default, Jun 27 2018, 14:40:03)
Nov 09 22:30:16 electrumx python3.7[548]: [GCC 8.1.0]
Nov 09 22:30:16 electrumx python3.7[548]: INFO:Controller:software version: ElectrumX 1.8.11
Nov 09 22:30:16 electrumx python3.7[548]: INFO:Controller:aiorpcX version: 0.10.1
Nov 09 22:30:16 electrumx python3.7[548]: INFO:Controller:supported protocol versions: 1.2-1.4
Nov 09 22:30:16 electrumx python3.7[548]: INFO:Controller:event loop policy: None
Nov 09 22:30:16 electrumx python3.7[548]: INFO:Controller:reorg limit is 2,000 blocks
Nov 09 22:30:16 electrumx python3.7[548]: INFO:Daemon:daemon #1 at 192.168.21.23:19337/ (current)
Nov 09 22:30:16 electrumx python3.7[548]: INFO:DB:switching current directory to /home/rleblanc/electrumx_db/ftc/
Nov 09 22:30:16 electrumx python3.7[548]: INFO:DB:using leveldb for DB backend
Nov 09 22:30:16 electrumx python3.7[548]: INFO:DB:opened UTXO DB (for sync: True)
Nov 09 22:30:16 electrumx python3.7[548]: INFO:DB:DB version: 6
Nov 09 22:30:16 electrumx python3.7[548]: INFO:DB:coin: Feathercoin
Nov 09 22:30:16 electrumx python3.7[548]: INFO:DB:network: mainnet
Nov 09 22:30:16 electrumx python3.7[548]: INFO:DB:height: 2,406,365
Nov 09 22:30:16 electrumx python3.7[548]: INFO:DB:tip: 430690c2ec2b73793b47b6a58e0e4690493f21870699944c95dd827f403e7121
Nov 09 22:30:16 electrumx python3.7[548]: INFO:DB:tx count: 4,605,127
Nov 09 22:30:16 electrumx python3.7[548]: INFO:DB:flushing DB cache at 1,200 MB
Nov 09 22:30:16 electrumx python3.7[548]: INFO:History:history DB version: 0
Nov 09 22:30:16 electrumx python3.7[548]: INFO:History:flush count: 34,373
Nov 09 22:30:16 electrumx python3.7[548]: ERROR:root:task crashed: <Task finished coro=<Controller.serve() done, defined at /usr/local/lib/python3.7/dist-packages/electrumX-1.8.11-py3.7.egg/electrumx/server/controller.py:81> exception=AssertionError()
Nov 09 22:30:16 electrumx python3.7[548]: Traceback (most recent call last):
Nov 09 22:30:16 electrumx python3.7[548]:   File "/usr/local/lib/python3.7/dist-packages/aiorpcX-0.10.1-py3.7.egg/aiorpcx/util.py", line 118, in check_task
Nov 09 22:30:16 electrumx python3.7[548]:     task.result()
Nov 09 22:30:16 electrumx python3.7[548]:   File "/usr/local/lib/python3.7/dist-packages/electrumX-1.8.11-py3.7.egg/electrumx/server/controller.py", line 131, in serve
Nov 09 22:30:16 electrumx python3.7[548]:     await group.spawn(wait_for_catchup())
Nov 09 22:30:16 electrumx python3.7[548]:   File "/usr/local/lib/python3.7/dist-packages/aiorpcX-0.10.1-py3.7.egg/aiorpcx/curio.py", line 242, in __aexit__
Nov 09 22:30:16 electrumx python3.7[548]:     await self.join(wait=self._wait)
Nov 09 22:30:16 electrumx python3.7[548]:   File "/usr/local/lib/python3.7/dist-packages/aiorpcX-0.10.1-py3.7.egg/aiorpcx/curio.py", line 213, in join
Nov 09 22:30:16 electrumx python3.7[548]:     raise task.exception()
Nov 09 22:30:16 electrumx python3.7[548]:   File "/usr/local/lib/python3.7/dist-packages/electrumX-1.8.11-py3.7.egg/electrumx/server/block_processor.py", line 654, in fetch_and_process_blocks
Nov 09 22:30:16 electrumx python3.7[548]:     await self._first_open_dbs()
Nov 09 22:30:16 electrumx python3.7[548]:   File "/usr/local/lib/python3.7/dist-packages/electrumX-1.8.11-py3.7.egg/electrumx/server/block_processor.py", line 634, in _first_open_dbs
Nov 09 22:30:16 electrumx python3.7[548]:     await self.db.open_for_sync()
Nov 09 22:30:16 electrumx python3.7[548]:   File "/usr/local/lib/python3.7/dist-packages/electrumX-1.8.11-py3.7.egg/electrumx/server/db.py", line 146, in open_for_sync
Nov 09 22:30:16 electrumx python3.7[548]:     await self._open_dbs(True, False)
Nov 09 22:30:16 electrumx python3.7[548]:   File "/usr/local/lib/python3.7/dist-packages/electrumX-1.8.11-py3.7.egg/electrumx/server/db.py", line 134, in _open_dbs
Nov 09 22:30:16 electrumx python3.7[548]:     await self._read_tx_counts()
Nov 09 22:30:16 electrumx python3.7[548]:   File "/usr/local/lib/python3.7/dist-packages/electrumX-1.8.11-py3.7.egg/electrumx/server/db.py", line 102, in _read_tx_counts
Nov 09 22:30:16 electrumx python3.7[548]:     assert len(tx_counts) == size
Nov 09 22:30:16 electrumx python3.7[548]: AssertionError
Nov 09 22:39:17 electrumx systemd[1]: Stopping Electrumx FTC...
Nov 09 22:39:17 electrumx python3.7[548]: WARNING:Controller:received SIGTERM signal, initiating shutdown
Nov 09 22:39:17 electrumx python3.7[548]: INFO:Controller:shutting down
Nov 09 22:39:17 electrumx python3.7[548]: INFO:Controller:shutdown complete
Nov 09 22:39:17 electrumx systemd[1]: Stopped Electrumx FTC.
Nov 09 22:39:22 electrumx systemd[1]: electrumx_ftc.service: Failed to reset devices.list: Operation not permitted
Nov 09 22:39:22 electrumx systemd[1]: Started Electrumx FTC.
Nov 09 22:39:22 electrumx python3.7[696]: INFO:Controller:Python version: 3.7.0 (default, Jun 27 2018, 14:40:03)
Nov 09 22:39:22 electrumx python3.7[696]: [GCC 8.1.0]
Nov 09 22:39:22 electrumx python3.7[696]: INFO:Controller:software version: ElectrumX 1.8.11
Nov 09 22:39:22 electrumx python3.7[696]: INFO:Controller:aiorpcX version: 0.10.1
Nov 09 22:39:22 electrumx python3.7[696]: INFO:Controller:supported protocol versions: 1.2-1.4
Nov 09 22:39:22 electrumx python3.7[696]: INFO:Controller:event loop policy: None
Nov 09 22:39:22 electrumx python3.7[696]: INFO:Controller:reorg limit is 2,000 blocks
Nov 09 22:39:22 electrumx python3.7[696]: INFO:Daemon:daemon #1 at 192.168.21.23:19337/ (current)
Nov 09 22:39:22 electrumx python3.7[696]: INFO:DB:switching current directory to /home/rleblanc/electrumx_db/ftc/
Nov 09 22:39:22 electrumx python3.7[696]: INFO:DB:using leveldb for DB backend
Nov 09 22:39:22 electrumx python3.7[696]: INFO:DB:opened UTXO DB (for sync: True)
Nov 09 22:39:22 electrumx python3.7[696]: INFO:DB:DB version: 6
Nov 09 22:39:22 electrumx python3.7[696]: INFO:DB:coin: Feathercoin
Nov 09 22:39:22 electrumx python3.7[696]: INFO:DB:network: mainnet
Nov 09 22:39:22 electrumx python3.7[696]: INFO:DB:height: 2,406,365
Nov 09 22:39:22 electrumx python3.7[696]: INFO:DB:tip: 430690c2ec2b73793b47b6a58e0e4690493f21870699944c95dd827f403e7121
Nov 09 22:39:22 electrumx python3.7[696]: INFO:DB:tx count: 4,605,127
Nov 09 22:39:22 electrumx python3.7[696]: INFO:DB:flushing DB cache at 1,200 MB
Nov 09 22:39:22 electrumx python3.7[696]: INFO:History:history DB version: 0
Nov 09 22:39:22 electrumx python3.7[696]: INFO:History:flush count: 34,373
Nov 09 22:39:22 electrumx python3.7[696]: ERROR:root:task crashed: <Task finished coro=<Controller.serve() done, defined at /usr/local/lib/python3.7/dist-packages/electrumX-1.8.11-py3.7.egg/electrumx/server/controller.py:81> exception=AssertionError()
Nov 09 22:39:22 electrumx python3.7[696]: Traceback (most recent call last):
Nov 09 22:39:22 electrumx python3.7[696]:   File "/usr/local/lib/python3.7/dist-packages/aiorpcX-0.10.1-py3.7.egg/aiorpcx/util.py", line 118, in check_task
Nov 09 22:39:22 electrumx python3.7[696]:     task.result()
Nov 09 22:39:22 electrumx python3.7[696]:   File "/usr/local/lib/python3.7/dist-packages/electrumX-1.8.11-py3.7.egg/electrumx/server/controller.py", line 131, in serve
Nov 09 22:39:22 electrumx python3.7[696]:     await group.spawn(wait_for_catchup())
Nov 09 22:39:22 electrumx python3.7[696]:   File "/usr/local/lib/python3.7/dist-packages/aiorpcX-0.10.1-py3.7.egg/aiorpcx/curio.py", line 242, in __aexit__
Nov 09 22:39:22 electrumx python3.7[696]:     await self.join(wait=self._wait)
Nov 09 22:39:22 electrumx python3.7[696]:   File "/usr/local/lib/python3.7/dist-packages/aiorpcX-0.10.1-py3.7.egg/aiorpcx/curio.py", line 213, in join
Nov 09 22:39:22 electrumx python3.7[696]:     raise task.exception()
Nov 09 22:39:22 electrumx python3.7[696]:   File "/usr/local/lib/python3.7/dist-packages/electrumX-1.8.11-py3.7.egg/electrumx/server/block_processor.py", line 654, in fetch_and_process_blocks
Nov 09 22:39:22 electrumx python3.7[696]:     await self._first_open_dbs()
Nov 09 22:39:22 electrumx python3.7[696]:   File "/usr/local/lib/python3.7/dist-packages/electrumX-1.8.11-py3.7.egg/electrumx/server/block_processor.py", line 634, in _first_open_dbs
Nov 09 22:39:22 electrumx python3.7[696]:     await self.db.open_for_sync()
Nov 09 22:39:22 electrumx python3.7[696]:   File "/usr/local/lib/python3.7/dist-packages/electrumX-1.8.11-py3.7.egg/electrumx/server/db.py", line 146, in open_for_sync
Nov 09 22:39:22 electrumx python3.7[696]:     await self._open_dbs(True, False)
Nov 09 22:39:22 electrumx python3.7[696]:   File "/usr/local/lib/python3.7/dist-packages/electrumX-1.8.11-py3.7.egg/electrumx/server/db.py", line 134, in _open_dbs
Nov 09 22:39:22 electrumx python3.7[696]:     await self._read_tx_counts()
Nov 09 22:39:22 electrumx python3.7[696]:   File "/usr/local/lib/python3.7/dist-packages/electrumX-1.8.11-py3.7.egg/electrumx/server/db.py", line 102, in _read_tx_counts
Nov 09 22:39:22 electrumx python3.7[696]:     assert len(tx_counts) == size
Nov 09 22:39:22 electrumx python3.7[696]: AssertionError

**Comment by rldleblanc** [comment]
We had a "planned" power outage today. Now all my ElectrumX nodes are broken the same way (I recovered them all the other day).

**Comment by kyuupichan** [comment]
I think you'll find if you kill -9 ElectrumX at any point nothing gets corrupted.  So the fact that things corrupt when the power goes off indicates an O/S, hardware or filesystem issue.

**Comment by rldleblanc** [comment]
Any way to truncate up to some stable point, it do I have to start over
from Genesis?

I'm moving to a new power supply in the next day or two. I'll test yanking
the power on the new supply and see if I still get the same problem.

**Comment by kyuupichan** [comment]
If you do a "stop" rpc, or kill (not -9) the process, it will shut down cleanly.  Then you can make a backup of the DB.

I don't recommend yanking the power supply as that seems to corrupt the DB.  Try kill -9 the server process a few times and see if it starts up cleanly; if it does then the issue is outside ElectrumX

**Comment by kyuupichan** [comment]
Sure.  But you appreciate that if kill -9 works and pulling the power doesn't, the problem is not ElectrumX right?  Of course, improvements welcome if you find anything.

**Comment by rldleblanc** [comment]
Yes, I agree. Dirty shutdowns are not caused by the software. It's what is
in the buffers and how things are written to disk (all OSes reorder blocks
as they are written to disk). If ordering to disk is critical in an
application, then it is the application's responsibility to use barriers
and/or sync I/O. ElectrumX doesn't need that because it has Bitcoin that is
the authority.

**Comment by rldleblanc** [comment]
Hmm, I added some print commands and restarted one of the daemons and it started right up. I then tried others on the non-modified code and they started up just fine. I wonder if in this case it has something to do with the bitcoin daemon not being ready yet (the original case was restarted multiple times, but the damage could have been done with an older version already). I have my ElectrumX daemons in an LXC container and my coin daemons in different containers so I can't order them in SystemD. I'll poke around with it some more.
